docs/blog/posts/2024-12-11-ultralytics-attack-analysis.md

```diff
@@ -0,0 +1,68 @@
+
+The second round of malicious releases came from the attacker using an unrevoked PyPI API token that was still available to the GitHub Actions workflow, potentially a hold-over from before the project adopted Trusted Publishing. This was detectable because there were no corresponding source repository activity or PyPI publish attestations for the second round of releases.
+
+Once tools begin utilizing these publish attestations to record the “expected” provenance of software, this type of attack will be less effective as the lack of provenance information will be more apparent and verifiable at install time.
+
+## How can PyPI improve the security of the ecosystem?
+
+From this story, we can see a few places where PyPI can help developers towards a secure configuration without infringing on existing use-cases:
+
+* **API tokens are allowed to go unused alongside Trusted Publishers.** It’s valid for a project to use a mix of API tokens and Trusted Publishers because Trusted Publishers aren’t universally supported by all platforms. However, API tokens that are being unused over a period of time despite releases continuing to be published via Trusted Publishing is a strong indicator that the API token is no longer needed and can be revoked.
+
+* **GitHub Environments are optional, but recommended, when using a GitHub Trusted Publisher.** However, PyPI doesn’t fail or warn users that are using a GitHub Environment that the corresponding Trusted Publisher isn’t configured to require the GitHub Environment. This fact didn’t end up mattering for this specific attack, but during the investigation it was noticed as something easy for project maintainers to miss.
+
+We’ve created tracking issues for both of the above changes.
+
+
+## What is PyPI doing about malware?
+
+Not every package and release on PyPI should be treated as trusted, it is up to you the user to review your usage of software from PyPI before choosing to install packages.
+
+PyPI staff and volunteers do their best to remove malware, but because the service is open to anyone looking to publish software there is an unfortunately high amount of abuse. Thankfully most of this abuse does not have the same widespread impact as a targeted attack on an already widely-used project.
+
+Mike Fiedler, the PyPI Safety and Security Engineer is working on new systems for reducing the time that malware is available to be installed on PyPI, through [APIs that security researchers can automatically send reports](./2023-09-18-inbound-malware-reporting/) to and new [“quarantine” release status](./2024-08-16-safety-and-security-engineer-year-in-review/#project-lifecycle-status-quarantine) to prevent harm while a human investigates the situation. Expect more in this space in 2025!
```

Comment:
Still triggers INFO-level logs (aside, why doesn't mkdocs think those are WARN?)
change:
```suggestion
Mike Fiedler, the PyPI Safety and Security Engineer is working on new systems for reducing the time that malware is available to be installed on PyPI, through [APIs that security researchers can automatically send reports](./2023-09-18-inbound-malware-reporting.md) to and new [“quarantine” release status](./2024-08-16-safety-and-security-engineer-year-in-review.md#project-lifecycle-status-quarantine) to prevent harm while a human investigates the situation. Expect more in this space in 2025!
```